lib/FrontendTool/FrontendTool.cpp

```diff
@@ -941,16 +947,26 @@ static void emitReferenceDependenciesForAllPrimaryInputsIfNeeded(
         Invocation.getReferenceDependenciesFilePathForPrimary(
             SF->getFilename());
     if (!referenceDependenciesFilePath.empty()) {
-      if (Invocation.getLangOptions().EnableFineGrainedDependencies)
+      auto LangOpts = Invocation.getLangOptions();
+      ReferencedNameTracker *tracker = nullptr;
+      if (LangOpts.DisableRequestBasedIncrementalDependencies) {
+        tracker = SF->getReferencedNameTracker();
+      } else {
+        tracker = SF->getRequestBasedReferencedNameTracker();
+      }
+      if (LangOpts.EnableFineGrainedDependencies) {
         (void)fine_grained_dependencies::emitReferenceDependencies(
             Instance.getASTContext().Diags, SF,
-            *Instance.getDependencyTracker(), referenceDependenciesFilePath,
-            Invocation.getLangOptions()
-                .EmitFineGrainedDependencySourcefileDotFiles);
-      else
+            *Instance.getDependencyTracker(),
+            tracker,
+            referenceDependenciesFilePath,
+            LangOpts.EmitFineGrainedDependencySourcefileDotFiles);
+      } else {
         (void)emitReferenceDependencies(Instance.getASTContext().Diags, SF,
                                         *Instance.getDependencyTracker(),
+                                        tracker,
                                         referenceDependenciesFilePath);
+      }
```

Comment:
I also prefer (as mentioned elsewhere) to pass in the boolean that says if the new thing is enabled, rather than pass in a whole pointer to the tracker.